app/src/main/java/org/mozilla/fenix/home/HomeFragment.kt

```diff
@@ -923,30 +925,41 @@ class HomeFragment : Fragment() {
             ?.isVisible = tabCount > 0
     }
 
-    private fun displayWallpaperIfEnabled() {
+    private fun observeWallpaperChanges() {
         if (shouldEnableWallpaper()) {
             requireComponents.appStore.flow()
                 .ifChanged { state -> state.wallpaperState.currentWallpaper }
                 .onEach { state ->
-                    // We only want to update the wallpaper when it's different from the default one
-                    // as the default is applied already on xml by default.
-                    when (val currentWallpaper = state.wallpaperState.currentWallpaper) {
-                        Wallpaper.Default -> {
-                            binding.wallpaperImageView.visibility = View.GONE
-                        }
-                        else -> {
-                            val bitmap = requireComponents.useCases.wallpaperUseCases.loadBitmap(currentWallpaper)
-                            bitmap?.let {
-                                it.scaleToBottomOfView(binding.wallpaperImageView)
-                                binding.wallpaperImageView.visibility = View.VISIBLE
-                            }
-                        }
-                    }
+                    showWallpaper(state.wallpaperState.currentWallpaper)
                 }
                 .launchIn(viewLifecycleOwner.lifecycleScope)
         }
     }
 
+    private fun setWallpaperToCurrent() {
+        if (shouldEnableWallpaper()) {
+            val wallpaper = requireComponents.appStore.state.wallpaperState.currentWallpaper
+            runBlockingIncrement {
+                showWallpaper(wallpaper)
+            }
+        }
+    }
+
+    private suspend fun showWallpaper(wallpaper: Wallpaper) = when (wallpaper) {
```

Comment:
Since we are calling `setWallpaperToCurrent()` and `observeWallpaperChanges()`, is it possible for the same wallpaper to be shown twice? If so, can we add a check to not do anything if the wallpaper to be shown is already the currently shown wallpaper?

Reply:
> Since we are calling setWallpaperToCurrent() and observeWallpaperChanges()

There is some conversation above about whether it might even be correct for now to remove `observeWallpaperChanges`. Doing so could clean this up, but it might add extra work to the wallpaper selection tool. The wallpaper would indeed get shown twice using the current strategy, but there is no visual interruption as far as I can tell. It's possible the platform does something smart in terms of caching `ImageView`s, but I don't know for sure.

> If so, can we add a check to not do anything if the wallpaper to be shown is already the currently shown wallpaper?

I would be open to this strategy. I generally try to avoid introducing mutable state where possible, but this sounds like it might be the simplest solution.

Reply:
Ya, I would want to avoid introducing a mutable state as well. See https://github.com/mozilla-mobile/fenix/pull/26515#discussion_r950405365 if this could be a potential solution with move forward with to unblock this.